api/envoy/service/ext_proc/v3alpha/external_processor.proto

```diff
@@ -0,0 +1,300 @@
+  // based on the headers and other metadata that they see.
+  envoy.extensions.filters.http.ext_proc.v3alpha.ProcessingMode mode_override = 9;
+}
+
+// The following are messages that are sent to the server.
+
+// This message is sent to the external server when the HTTP request and responses
+// are first received.
+message HttpHeaders {
+  // The HTTP request headers. All header keys will be
+  // lower-cased, because HTTP header keys are case-insensitive.
+  config.core.v3.HeaderMap headers = 1;
+
+  // The values of properties selected by the "request_properties"
+  // list in the configuration. Each entry in the map is the
+  // value of the specified property.
+  map<string, google.protobuf.Struct> properties = 2;
+
+  // If true, then there is no message body associated with this
+  // request or response.
+  bool end_of_stream = 3;
+}
+
+// This message contains the message body that Envoy sends to the external server.
+message HttpBody {
+  bytes body = 1;
+
+  bool end_of_stream = 2;
+}
+
+// This message contains the trailers.
+message HttpTrailers {
+  config.core.v3.HeaderMap trailers = 1;
+}
+
+// The following are messages that may be sent back by the server.
+
+// This message must be sent in response to a request_headers or response_headers
+// message.
+// [#next-free-field: 6]
+message HeadersResponse {
+  // Instructions on how to manipulate the headers
+  HeaderMutation header_mutation = 1;
+
+  // If set, add a body to the message if none was present
+  // before.
+  BodyMutation body_mutation = 2;
+
+  // If set, add trailers to the message. If and only if this
+  // field is combined with the end_stream flag, then
+  // the trailers will be sent along with the request or
+  // response -- otherwise, this field is ignored.
+  config.core.v3.HeaderMap trailers = 3;
+
+  // If true, then replace the request and response with the contents
+  // of this message. If header_mutation is set, apply it to the
+  // headers. If body_mutation is set and contains a body, then add that
+  // body to the request or response, even if one does not already exist --
+  // otherwise, clear the body. Any additional body and trailers
+  // received from downstream will be ignored, and any additional messages
+  // from the remote server on either the request or response path will
+  // also be ignored for the duration of this request or response.
+  bool end_of_stream = 4;
+
+  // Clear the route cache for the current request.
+  // This is necessary if the remote server
+  // modified headers that are used to calculate the route.
+  bool clear_route_cache = 5;
+}
+
+// This message must be sent in response to a request_trailers
+// or response_trailers message.
+message TrailersResponse {
+  // Instructions on how to manipulate the trailers
+  HeaderMutation header_mutation = 1;
+}
+
+// TODO this is now the same as HeaderResponse. However, do we think
+// that will always be the case? Keeping it as a separate message
+// for now.
```

Comment:
You might consider a common shared message and a wrapper to allow for extensibility.

Reply:
I'll take a look later today or tomorrow.